janus.go

```diff
@@ -126,6 +254,88 @@ func (gateway *Gateway) sendloop() {
 
 }
 
+func (gateway *Gateway) recvHttpResp(resp []byte) error {
+	// Decode to Msg struct
+	var base BaseMsg
+	if err := json.Unmarshal(resp, &base); err != nil {
+		fmt.Printf("json.Unmarshal: %s\n", err)
+		return err
+	}
+
+	if debug {
+		// log message being sent
+		var log bytes.Buffer
+		json.Indent(&log, resp, "<", "   ")
+		log.Write([]byte("\n"))
+		log.WriteTo(os.Stdout)
+	}
+
+	typeFunc, ok := msgtypes[base.Type]
+	if !ok {
+		fmt.Printf("Unknown message type received!\n")
+		return errors.New("Unknown base type received")
+	}
+
+	msg := typeFunc()
+	if err := json.Unmarshal(resp, &msg); err != nil {
+		fmt.Printf("json.Unmarshal: %s\n", err)
+		return err
+	}
+
+	// Pass message on from here
+	if base.ID == "" {
+		// Is this a Handle event?
+		if base.Handle == 0 {
+			// Error()
+		} else if base.Type == "trickle" {
+			// Lookup Session
+			gateway.Lock()
+			session := gateway.Sessions[base.Session]
+			gateway.Unlock()
+			if session == nil {
+				fmt.Printf("Unable to deliver message. Session gone?\n")
+				return errors.New("Unable to deliver message")
+			}
+			// Pass msg to session
+			go passMsg(session.events, msg)
+		} else {
+			// Lookup Session
+			gateway.Lock()
```

Comment:
I'm not sure what the pressure on this service is likely to be like; is it worth making this a RWMutex and using `RLock` here?

Reply:
This is from their original code from the ws recv, the gateway is going to be used per lambda session so it should be ok but if you feel strongly about it I can change it

Reply:
Nah it's probably fine, just depends how often the map is read from and whether there's a performance hit if you exclude mutual reads. But if there aren't many concurrent reads it should be fine.